File: java/Median/src/Median.java
/* 
Даны два отсортированных числовых массива одинаковой длины N. 
Найдите медиану числового массива длины 2N, содержащего все 
числа из двух данных массивов.
*/

import java.util.ArrayList;
import java.util.Arrays;

public class Median {
	public static void main (String[] args) {
		ArrayList<Integer> listOne = new ArrayList<Integer>();
		listOne.addAll(Arrays.asList(new Integer[]{1, 2, 3, 4}));
		
		ArrayList<Integer> listTwo = new ArrayList<Integer>();
		listTwo.addAll(Arrays.asList(new Integer[]{1, 4, 5, 6}));

		Service serv = new Service();
		System.out.println(serv.getMedian(listOne, listTwo)); // => 3.5
	}
	
}

class Service {
	public double getMedian(ArrayList<Integer> listOne, ArrayList<Integer> listTwo) {
		int halfLength = listOne.size();
		ArrayList<Integer> resultList = this.merge(listOne, listTwo); // слияние отсортированных массивов

		return (resultList.get(halfLength - 1) + resultList.get(halfLength)) / 2.0;
	}

	private ArrayList<Integer> merge(ArrayList<Integer> left, ArrayList<Integer> right) {
		ArrayList<Integer> result = new ArrayList<Integer>();
		while(left.size() > 0 && right.size() > 0) {
			if(left.get(0) <= right.get(0)) {
				result.add(left.remove(0));
			}
			else {
				result.add(right.remove(0));
			}
		}

		if(left.size() > 0) {
			result.addAll(left);
		}

		if(right.size() > 0) {
			result.addAll(right);
		}

		return result;
	}
}

File: python/median.py
# -*- coding: utf-8 -*-
u"""
Даны два отсортированных числовых массива одинаковой длины N. 
Найдите медиану числового массива длины 2N, содержащего все 
числа из двух данных массивов.
"""

def merge(left, right):
    result = []
    while len(left) > 0 and len(right) > 0:
        if left[0] <= right[0]:
            result.append(left.pop(0))
        else:
            result.append(right.pop(0))

    if len(left) > 0:
        result += left

    if len(right) > 0:
        result += right 

    return result
    

def get_median(list_one, list_two):
    half_length = len(list_one)
    result_list = merge(list_one, list_two)  # слияние отсортированных массивов

    return (result_list[half_length - 1] + result_list[half_length]) / 2.0


print get_median([1, 2, 3, 4], [1, 4, 5, 6])  # => 3.5